Support linking mcasts within/across subcmds for kernel bins

tt-aho opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/11014

### Problem description
We took a performance hit due to 35d7055, as we were previously never reserving our mcast path in dispatch, leading to potential deadlock. This pr enabling the linking option for mcast to minimize required reservations/VC allocations to recover performance.

### What's changed
Changed cb_acquire_pages to use a free running counter instead of issuing atomic inc. This allows us to enable linking within subcmds and removes unnecessary atomic txns/barriers.
Changed dispatch upstream communication to use the alternative noc. The is an rtl issue where no other txns can be issued when linking is enabled, even if they use a different command buffer/VC. This enables linking across subcmds.

### Checklist
- [x] Post commit CI passes
- [x] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

tt-aho committed:
#11014: Add a flag to CQDispatchWritePackedLargeSubCmd to specify unlinking mcasts
Update assemble_device_commands to set unlink flag when changing the mcast grid
Refactor write_packed_large to enable setting linked flag on first txn, and unset on last txn to a given mcast grid

tt-aho committed:
#11014: Change cb_acquire_pages to be a free running counter to enable linking within a subcmd

tt-aho committed:
#11014: Use NOC 1 for dispatch upstream communication to support linking across subcmds

tt-aho committed:
#11014: Enable linking within and across subcmds

pgkeller commented on tt_metal/impl/dispatch/cq_commands.hpp:
```diff
@@ -178,7 +178,6 @@ struct CQDispatchWritePagedCmd {
 constexpr uint32_t CQ_DISPATCH_CMD_PACKED_WRITE_FLAG_NONE      = 0x00;
 constexpr uint32_t CQ_DISPATCH_CMD_PACKED_WRITE_FLAG_MCAST     = 0x01;
 constexpr uint32_t CQ_DISPATCH_CMD_PACKED_WRITE_FLAG_NO_STRIDE = 0x02;
-constexpr uint32_t CQ_DISPATCH_CMD_PACKED_WRITE_RELAY = 0x03;
```
pedantic: spacing doesn't match

tt-aho replied:
This is actually deleting the line that doesn't match spacing 🙂

pgkeller commented on tt_metal/impl/dispatch/cq_commands.hpp:
```diff
@@ -197,11 +196,16 @@ struct CQDispatchWritePackedMulticastSubCmd {
     uint32_t addr;
     uint16_t length;          // multiples of L1 cache line alignment
     uint16_t num_mcast_dests;
+    uint8_t flags;
```
I think num_mcast_dsts can be a uint8_t and we can pack flags in, right?

tt-aho replied:
Ah yes, we have it as uint8_t in some other places too so should be fine to change.

pgkeller reviewed (commented):
these changes have huge implications, may want to run through the stress test CI suite before merging

tt-aho committed:
#0: Update all noc sanitization apis to take in noc index for validation instead of assuming default noc

tt-aho commented on tt_metal/impl/dispatch/kernels/cq_common.hpp:
```diff
@@ -217,25 +217,23 @@ uint32_t cb_acquire_pages(uint32_t cb_fence,
-        // Ensure last sem_inc has landed
-        noc_async_atomic_barrier();
+    static uint32_t counter = 0;
+    static uint32_t sem = 0;
```
I can rename, though wanted to avoid having upstream as a variable name in this function since it seemed generic/not necessarily upstream. Would remote_count be better?

tt-aho committed:
#11014: Move local counter from being a static variable of cb_acquire_pages and make is a provided arg.
cq_prefetch and cq_dispatch create their own global to be passed in for the count

tt-aho commented on tt_metal/impl/dispatch/kernels/cq_dispatch.cpp:
```diff
@@ -1094,7 +1095,8 @@ void kernel_main() {
     // Confirm expected number of pages, spinning here is a leak
-    cb_wait_all_pages<my_dispatch_cb_sem_id>(0);
+    // TODO: We need to pass in our static counter here
+    // cb_wait_all_pages<my_dispatch_cb_sem_id>(0);
```
I refactored it to be global rather than static to the fn. We pass the count value into here now

tt-aho committed:
#0: Comment out some dispatcher dprints to fit on GS

tt-aho committed:
#0: Make WatcherFixture always set NoInline to avoid code size overflow

abhullar-tt commented on tt_metal/impl/device/device.cpp:
```diff
@@ -1342,6 +1345,10 @@ void Device::compile_command_queue_programs() {
     constexpr uint32_t prefetch_d_downstream_cb_sem = 2;
     constexpr uint32_t dispatch_downstream_cb_sem = 1;
 
+    // TODO: this->hw_command_queues_[cq_id]->noc_index is also hardcoded to NOC_0 elsewhere, should have one definition
```
does this need to be checked with some assert?

tt-aho replied:
Yeah I'll add an assert for this + the downstream noc must be different from the default noc.
